File: Julia/KerasNN/KerasNN.jl
using CSV, DataFrames, Dates, Flux, Plots

# Charger les données à partir du fichier CSV
data = CSV.read("Dataset/KerasNN/airPassengers.csv", DataFrame)

# Convert 'month' column in Date type
data.month = Dates.Date.(string.(data.month), "yyyy-mm-dd")

# Extract year and month as numerical features
data.year = Dates.year.(data.month)
data.month = Dates.month.(data.month)

# Replace months by a sequence of numbers from 1 to the length of the months column
data.month = 1:length(data.month)

# Extracting time sequences
sequences = [Float32.(data.passengers)]

# Splitting training data and test data
train_size = floor(Int, size(data, 1) * 0.8)
train_data = data[1:train_size, :]
test_data = data[train_size+1:end, :]

# Building a model with 2 dense layers
model = Chain(
    Dense(1, 10, relu),
    Dense(10, 1)
)

# Prepare training data
X_train = transpose(Float32.(train_data.month))
Y_train = transpose(Float32.(train_data.passengers))

X_test = transpose(Float32.(test_data.month))
Y_test = transpose(Float32.(test_data.passengers))

# Defining loss function and optimizer
loss(x, y) = Flux.mse(model(x), y)
optimizer = ADAM()

# Training the model
Flux.train!(loss, Flux.params(model), [(X_train, Y_train)], optimizer)

# Model evaluation
test_loss = loss(X_test, Y_test)
println("Test Loss: $test_loss")

# Make predictions
predictions = model(X_test)

# Convert Y_test and predictions to vectors
Y_test_vec = vec(Y_test)
predictions_vec = vec(predictions)

# Tracing predictions against actual data
plot(test_data.month, Y_test_vec, label="Actual")
plot!(test_data.month, predictions_vec, label="Predicted")

File: Python/KerasNN/KerasNN.py
import pandas as pd
import numpy as np
from sklearn.preprocessing import MinMaxScaler

# The following 2 lines may show an error, but the venv is ignoring the errors
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

import matplotlib.pyplot as plt

# Loading the dataset
file_path = "../../Dataset/Arima/airPassengers.csv"
df = pd.read_csv(file_path)
df["month"] = pd.to_datetime(df["month"])
df = df.set_index("month")

# Processing the Data
scaler = MinMaxScaler()
df_scaled = scaler.fit_transform(df)

# Splitting the data into training and testing sets
train_size = int(len(df_scaled) * 0.8)
train_data, test_data = df_scaled[:train_size], df_scaled[train_size:]


# Function to create sequences for time series data training
def create_sequences(data, seq_length):
    x, y = [], []
    for i in range(len(data) - seq_length):
        seq_in = data[i:i + seq_length]
        seq_out = data[i + seq_length]
        x.append(seq_in)
        y.append(seq_out)
    return np.array(x), np.array(y)


# Define the sequence length
sequence_length = 12

# Creating sequences for training data
x_train, y_train = create_sequences(train_data, sequence_length)

# Creating a simple feedforward neural network model
model = Sequential()
model.add(Dense(50, activation="relu", input_dim=sequence_length))
model.add(Dense(1))
model.compile(optimizer="adam", loss="mean_squared_error")

# Training the model
model.fit(x_train, y_train, epochs=100, batch_size=1)

# Testing the model on the test data
test_inputs = df_scaled[len(df_scaled) - len(test_data) - sequence_length:]
test_inputs = test_inputs.reshape((1, -1))

x_test = []

for i in range(sequence_length, len(test_inputs[0])):
    x_test.append(test_inputs[0, i - sequence_length:i])

x_test = np.array(x_test)

# Making predictions
predictions = model.predict(x_test)
predictions = scaler.inverse_transform(predictions)

# Plotting the results
train = df[:train_size]
test = df[train_size:].copy()
test.loc[:, "Predictions"] = predictions

plt.figure(figsize=(12, 6))
plt.plot(train["passengers"], label="Training Data")
plt.plot(test["passengers"], label="Testing Data")
plt.plot(test["Predictions"], label="Predictions")
plt.title("Air Passengers Time Series Prediction")
plt.xlabel("Year")
plt.ylabel("Passengers")
plt.legend()
plt.show()
